### ISSUE TITLE
Truncated Gamma

### ISSUE THREAD

**Issue by quattro** [pyro-ppl/numpyro#969]
It would be great to have a [truncated gamma distribution](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3898536/) implemented in NumPyro in order to cover lower-bounded variances (or upper bounded precisions) in probabilistic programs. 

I'm happy to code this up and issue a pull-req after some testing.

A nice example would be in Mendelian Randomization from summary statistics (ie two-stage least squares). Residual variance greater than 1 is consistent with heterogeneity across studies, but inferred variance should be bounded below by 1.

**Comment by fehiepsi** [comment]
@quattro I didn't use Truncated Gamma previously but I think this would be nice to have. FYI, we have [TruncatedDistribution](https://github.com/pyro-ppl/numpyro/blob/7493e35ed921108d884a88f58ab92ec5565f845d/numpyro/distributions/continuous.py#L1335) API with base distributions can be Cauchy, Laplace, Logistic, Normal, or StudentT. Probably it would be cleaner to follow that setting and have LeftTruncatedGamma, RightTruncatedGamma, TwoSidedTruncatedGamma implemented, then dispatching `TruncatedGamma` to the corresponding class (based on `low=None` or `high=None`).

**Comment by quattro** [comment]
Great, thanks for the advice @fehiepsi. I'll likely be busy with some other things for a little while but plan on coming back to this soon.

**Comment by quattro** [comment]
Hi @fehiepsi , I made some progress on this, but stopped at the sampling implementation. I think there are two paths forward, and would appreciate your thoughts on the matter:

1. Use a `uniform` -> `invCDF(gamma)` sampling based approach
2. Use rejection sampling

In terms of implementation:
1. Requires `gammaincinv` in order to compute the quantiles of the `gamma` distribution, which is currently not supported by JAX (as you pointed out here: https://github.com/google/jax/issues/5350). It looks like TFP has `gammaincinv` implemented (along with gradients), but admittedly, I don't have the bandwidth to initiate a PR and port it over to JAX.
2. There are a few papers describing sampling approaches to either the [Left/Right truncated Gamma](https://link.springer.com/article/10.1023/A:1018534102043), or the [more general case using latent variables](https://www.jstor.org/stable/1391008), that might be worthwhile investigating.

I'm hoping to move forward with this soon, as it looks like there is other support for a truncated gamma distribution, as mentioned in this thread: https://github.com/google/jax/issues/552 .

**Comment by fehiepsi** [comment]
@quattro I think the simplest solution is to make a wrapper (with try/except import like [contrib.tfp](https://github.com/pyro-ppl/numpyro/blob/master/numpyro/contrib/tfp/__init__.py)) for `gammaincinv` in `numpyro.distributions.util` and use it in the sample method. That would unblock you. What do you think?

**Comment by fehiepsi** [comment]
Sure, @quattro! We can discuss the numerical issue in detail in the PR. :)

**Comment by deeb01** [comment]
The `gammaincinv` blocker in this thread is resolved: the wrapper @fehiepsi suggested now
exists as `numpyro.distributions.util.gammaincinv`, and `Gamma` has both `cdf` and `icdf`
on master, so inverse-CDF sampling needs no new numerics.

@quattro — #1187 has been inactive since 2022. Are you still planning to finish it? If
not, I'd like to open a fresh PR following the `LeftTruncatedGamma` /
`RightTruncatedGamma` / `TwoSidedTruncatedGamma` structure @fehiepsi outlined above, with
credit to your branch.

**Comment by Qazalbash** [comment]
> The `gammaincinv` blocker in this thread is resolved: the wrapper [@fehiepsi](https://github.com/fehiepsi) suggested now exists as `numpyro.distributions.util.gammaincinv`, and `Gamma` has both `cdf` and `icdf` on master, so inverse-CDF sampling needs no new numerics.
> 
> [@quattro](https://github.com/quattro) — [#1187](https://github.com/pyro-ppl/numpyro/pull/1187) has been inactive since 2022. Are you still planning to finish it? If not, I'd like to open a fresh PR following the `LeftTruncatedGamma` / `RightTruncatedGamma` / `TwoSidedTruncatedGamma` structure [@fehiepsi](https://github.com/fehiepsi) outlined above, with credit to your branch.

Hi @deeb01, I believe you are right; it's been inactive for many years. I was also thinking of taking charge of it, but please feel free to open a PR.
